src/Interpreters/Context.cpp

```diff
@@ -1816,6 +1816,57 @@ std::optional<UInt16> Context::getTCPPortSecure() const
     return {};
 }
 
+// Small compile time FNV-1a 64 hashing.
+constexpr size_t compileTimeHashing(const std::string_view & s)
+{
+    size_t res = 0xcbf29ce484222325;
+    for (auto c : s)
+    {
+        res ^= c;
+        res *= 1099511628211;
+    }
+    return res;
+}
+
+constexpr size_t operator"" _h(const char * p, size_t s)
+{
+    return compileTimeHashing(std::string_view(p, s));
+}
+
+UInt16 Context::getPort(const std::string_view & port_name) const
+{
+    auto h = compileTimeHashing(port_name);
+    auto lock = getLock();
+    const auto & config = getConfigRef();
+    switch (h)
+    {
+        case "tcp"_h:
+            return config.getInt("tcp_port", DBMS_DEFAULT_PORT);
+        case "tcp_with_proxy"_h:
+            return config.getInt("tcp_with_proxy_port", 0);
+        case "tcp_secure"_h:
+            return config.getInt("tcp_port_secure", DBMS_DEFAULT_SECURE_PORT);
+        case "http"_h:
+            return config.getInt("http_port", DBMS_DEFAULT_HTTP_PORT);
+        case "https"_h:
+            return config.getInt("http_port", 0);
+        case "interserver_http"_h:
+            return config.getInt("interserver_http_port", 0);
+        case "interserver_https"_h:
+            return config.getInt("interserver_https_port", 0);
+        case "postgresql"_h:
+            return config.getInt("postgresql_port", 0);
+        case "mysql"_h:
+            return config.getInt("mysql_port", 0);
+        case "grpc"_h:
+            return config.getInt("grpc_port", 0);
+        case "prometheus"_h:
```

Comment:
I'm afraid we will always forget to make changes here when adding new server interfaces.

Also this PR can be inconsistent with #27442 (I did not read it though).

Let's think how to restructure the code.

Reply:
Indeed. I think we should hold the server port registery in global context. I'll refactor.